File: cpp/SlidingWindow/findAnagrams/findAnagrams.cpp
#include <string>
#include <vector>
class Solution {
  public:
    std::vector<int> findAnagrams(std::string s, std::string p) {
        std::vector<int> res = {};
        std::vector<int> pCount(26, 0);
        std::vector<int> windowCount(26, 0);

        int pLen = p.size();
        int sLen = s.size();

        for (int i = 0; i < pLen; i++) {
            pCount[p[i] - 'a']++;
            windowCount[s[i] - 'a']++;
        }

        if (pCount == windowCount) {
            res.push_back(0);
        }

        for (int right = pLen; right < sLen - pLen; right++) {
            int left = right - pLen;
            windowCount[s[right] - 'a']++;
            windowCount[s[left] - 'a']--;

            if (pCount == windowCount) {
                res.push_back(left + 1);
            }
        }

        return res;
    }
};

File: java/SlidingWindow/findAnagrams/findAnagrams.java
import java.util.List;
import java.util.ArrayList;

class findAnagrams {
    public List<Integer> findAnagramsFunc(String s, String p) {

        int sLen = s.length();
        int pLen = p.length();

        if (sLen < pLen) {
            return new ArrayList<>();
        }

        List<Integer> res = new ArrayList<>();
        int[] pCount = new int[26];
        int[] windowCount = new int[26];

        for (int i = 0; i < pLen; i++) {
            pCount[p.charAt(i) - 'a']++;
            windowCount[s.charAt(i) - 'a']++;
        }

        if (pCount == windowCount) {
            res.add(0);
        }

        for (int right = pLen; right < sLen; right++) {
            int left = right - pLen;
            windowCount[s.charAt(left) - 'a']--;
            windowCount[s.charAt(right) - 'a']++;

            if (pCount == windowCount) {
                res.add(left + 1);
            }
        }
        return res;
    }
}
